projects/gloo/cli/pkg/cmd/install/knative.go

```diff
@@ -0,0 +1,320 @@
+				}
+			}
+			return nil
+		},
+	}
+	pflags := cmd.PersistentFlags()
+	flagutils.AddInstallFlags(pflags, &opts.Install)
+	flagutils.AddKnativeInstallFlags(pflags, &opts.Install.Knative)
+	return cmd
+}
+
+func installKnativeServing(opts *options.Options) error {
+	knativeOpts := opts.Install.Knative
+
+	// store the opts as a label on the knative-serving namespace
+	// we can use this to uninstall later on
+	knativeOptsJson, err := json.Marshal(knativeOpts)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := RenderKnativeManifests(knativeOpts)
+	if err != nil {
+		return err
+	}
+	if opts.Install.DryRun {
+		fmt.Printf("%s", manifests)
+		// For safety, print a YAML separator so multiple invocations of this function will produce valid output
+		fmt.Printf(yamlJoiner)
+		return nil
+	}
+
+	// install crds first by using selector
+	// https://knative.dev/docs/install/knative-with-any-k8s/
+	fmt.Fprintln(os.Stderr, "installing Knative CRDs...")
+	if err := install.KubectlApply([]byte(manifests), "--selector", knativeCrdSelector); err != nil {
+		return errors.Wrapf(err, "installing knative crds with kubectl apply")
+	}
+
+	fmt.Fprintln(os.Stderr, "installing Knative...")
+	if err := install.KubectlApply([]byte(manifests)); err != nil {
+		return errors.Wrapf(err, "installing knative crds with kubectl apply")
+	}
+	// label the knative-serving namespace as belonging to us
+	if err := install.Kubectl(nil, "annotate", "namespace",
+		"knative-serving", installedByUsAnnotationKey+"="+string(knativeOptsJson)); err != nil {
+		return errors.Wrapf(err, "annotating installation namespace")
+	}
+
+	fmt.Fprintln(os.Stderr, "Knative Successfully installed!")
+	return nil
+}
+
+func checkKnativeInstallation(kubeclient ...kubernetes.Interface) (bool, *options.Knative, error) {
+	var kc kubernetes.Interface
+	if len(kubeclient) > 0 {
+		kc = kubeclient[0]
+	} else {
+		kc = helpers.MustKubeClient()
+	}
+	namespaces, err := kc.CoreV1().Namespaces().List(v1.ListOptions{})
+	if err != nil {
+		return false, nil, err
+	}
+	for _, ns := range namespaces.Items {
+		if ns.Name == constants.KnativeServingNamespace {
+			if ns.Annotations != nil && ns.Annotations[installedByUsAnnotationKey] != "" {
+				installOpts := ns.Annotations[installedByUsAnnotationKey]
+				var opts options.Knative
+				if err := yaml.Unmarshal([]byte(installOpts), &opts); err != nil {
+					return false, nil, errors.Wrapf(err, "parsing install opts "+
+						"from knative-serving namespace annotation %v", installedByUsAnnotationKey)
+				}
+				return true, &opts, nil
+			}
+			return true, nil, nil
+		}
+	}
+	return false, nil, nil
+}
+
+// used by e2e test
+func RenderKnativeManifests(opts options.Knative) (string, error) {
+	knativeVersion := opts.InstallKnativeVersion
+	build := opts.InstallKnativeBuild
+	eventing := opts.InstallKnativeEventing
+	monitoring := opts.InstallKnativeMonitoring
+
+	servingRelease := fmt.Sprintf(servingReleaseUrl, knativeVersion)
```

Comment:
As per comment above, I think this would be clearer:
```suggestion
	servingReleaseUrl := fmt.Sprintf(servingReleaseUrlTemplate, knativeVersion)
```